### DISCUSSION TITLE
Documentation for VS Code Architecture/Anatomy?

### DISCUSSION THREAD

**Discussion by ericdrobinson** [microsoft/vscode-discussions#22] [Extension Development QnA] [answered]
As a VS Code extension developer I occasionally find the documentation insufficient to my needs or wonder if there isn’t some way that I can contribute/improve a feature. At present, I do not have a good way to understand VS Code’s architecture and find the relevant repository/location in code for me to investigate, review existing GitHub Issues, or open a new Issue.

What I'm interested in seeing is a birds eye view of "The Architecture of VS Code" (or perhaps "The Anatomy of VS Code"). How is VS Code put together? What code lives where? The [main VS Code repository](https://github.com/microsoft/vscode) appears to be a monorepo, though it does not include everything (e.g. the [Language Server](https://github.com/microsoft/vscode-languageserver-node) or [Debug Adapter](https://github.com/microsoft/vscode-debugadapter-node)). How do all of these pieces fit together? If I'm an extension developer working on a [debugger extension](https://code.visualstudio.com/api/extension-guides/debugger-extension), what are the repositories that are pertinent to me? What are the classes in the main codebase that my extension will interact with? Where in the main VS Code codebase can I look to see how the pieces fit together (in the event that the documentation doesn't contain some bit of information I need)?

I should point out that the [Related Projects wiki page](https://github.com/microsoft/vscode/wiki/Related-Projects) is close to what I'm interested in, though it appears to have the following issues:

1. It is incomplete and/or stale.
2. It lacks consistent organization.
3. It lacks any explanation about how the related projects are integrated into and used within VS Code.

At the end of every set of Release Notes is a fantastic list of "contributions to `vscode`" that include the specific projects affected. From [the most recent notes](https://code.visualstudio.com/updates/v1_67#_pull-requests) we see the following repositories:

- `vscode`
- `vscode-css-languageservice`
- `vscode-generator-code`
- `vscode-js-debug`
- `vscode-pull-request-github`
- `debug-adapter-protocol`
- `language-server-protocol`
- `monaco-editor`

The documentation I'm interested in finding would contain the above in some organization that allows me to understand how they fit together. (I would expect those projects to also, of course, be linked to their relevant repositories.) This documentation could even then be linked to from the Release Notes such that people who review them and find themselves inspired can find some quick way to zero-in on their interest (as well as understand how things fit together).

If something like this already exists, I'd love to be pointed towards it! 😄

**Comment by isidorn** [comment, marked as answer]
@ericdrobinson thanks for the great feedback. I believe no such documentation exists, and the road ahead should be to improve the [Related Projects wiki page](https://github.com/microsoft/vscode/wiki/Related-Projects).

fyi @gregvanl - might you know who owns the related projects page. I would assume we accept contributions to improve it, right?

**Reply by ericdrobinson** [reply]
> the road ahead should be to improve the [Related Projects wiki page](https://github.com/microsoft/vscode/wiki/Related-Projects).

Perhaps part of that road might include:

- Renaming the page (I would not necessarily think to look for information about VS Code's architecture/anatomy in a page simply listing related projects with links).
- A discussion (or at least a link) to the wiki page from the [extension docs](https://code.visualstudio.com/api). Having at least a brief introduction or pointer to a resource to "learn more if you need to go deeper" would be helpful.

?

**Reply by isidorn** [reply]
That sounds reasonable.

**Comment by robconery** [comment]
Closing for now.
